### PR TITLE
Glyphs 3, attempt 3

### PR DESCRIPTION
**Pull request by simoncozens**

### CODE REVIEW

File: Lib/glyphsLib/classes.py [hunk 1]
```diff
@@ -3383,9 +3472,12 @@ class GSGlyph(GSBase):
     )
 
     _classesForName = {
+        ".appVersion": str,
```
Comment (simoncozens) [inline Lib/glyphsLib/classes.py:3475]: These shouldn't be here, my git client went mad.

**Comment by simoncozens** [comment]
There's obviously now a lot of duplication between the `add_parser` lines and the `_classesForName`. But I will start working on the writer end soon, at which points `_classesForName` and `_keyOrder` will go away.

**Comment by anthrotype** [comment]
wait, are you no longer using openstep-plist? That'd be a bummer.. It's super fast, and I made it for this very purpose.

**Comment by simoncozens** [comment]
Yeah, not until this passes:

```
def test_roundtrip():
    infile = "tests/data/GlyphsFileFormatv3.glyphs"

    with open(infile, "r", encoding="utf-8") as fp:
        glyphs = fp.read()
        data = openstep_plist.loads(glyphs, use_numbers=True)
        out = openstep_plist.dumps(data, indent=0)
        assert glyphs == out
```

**Comment by simoncozens** [comment]
Third time's a charm, I think. I've made more progress in one afternoon than in the past week.

**Comment by anthrotype** [comment]
where can I find that file you mentioned "tests/data/GlyphsFileFormatv3.glyphs"?

**Comment by anthrotype** [comment]
I just tried to round-trip that "GlyphsFileFormatv3.glyphs" file of yours using `openstep_plist`, loading with `use_numbers=True`, then dumping and then re-loading and comparing the parsed and re-parsed data -- they match.
Is it the string formatting of the dumped .glyphs file that is the issue? E.g. white-space, indentation, etc.? I wouldn't worry too much about that. We could add a profile in the openstep_plist dump method that attempts to mimic the Glyphs.app formatting. Or we just don't care about matching it 1:1 if in order to do that the serializer has to have knowledge of the particular kind of object it is serializing.

**Comment by simoncozens** [comment]
I am planning to try and bring the OpenStep parser back in there later. I think it will be possible.

**Comment by anthrotype** [comment]
> bring the OpenStep parser back in there later

sounds good. It still isn't clear to me what exactly was the issue that you had earlier with openstep-plist, but I'll let you finish your work.

**Comment by simoncozens** [comment]
*Nearly* there now...

**Comment by simoncozens** [comment]
OK. This can roundtrip v2 and v3 files, *largely* into V3-style classes. However, it's not done yet. In order of priority:

- [ ] Translate ascender/xheight/etc. to/from new GSMetrics
- [ ] Translate customValue/widthValue/etc. to/from new GSAxes
- [ ] Test we can handle new-style color tuples
- [ ] See if we can tidy up the serialization routines in a more data-driven way
- [ ] See if we can use dictReader with openstep-plist alongside the existing Reader
- [ ] See if we can create a dictWriter with openstep-plist alongside the existing Writer

**Comment by simoncozens** [comment]
I've done what I can here. I have one failing test. (Duplicated because it uses defcon and ufoLib2.) I don't know why it's failing. The axis builder stuff is kinda beyond me.

**Comment by simoncozens** [comment]
Lots of bad things happening when we try to process it with fontmake:

```
WARNING:glyphsLib.builder.axes:Instance user-space location (400.0) redefined by 'Extra Light'
WARNING:glyphsLib.builder.axes:Instance user-space location (400.0) redefined by 'Light'
WARNING:glyphsLib.builder.axes:Instance user-space location (400) redefined by 'Regular'
WARNING:glyphsLib.builder.axes:Instance user-space location (400.0) redefined by 'Medium'
WARNING:glyphsLib.builder.axes:Instance user-space location (400.0) redefined by 'Bold'
WARNING:glyphsLib.builder.axes:Instance user-space location (400.0) redefined by 'Black'
WARNING:glyphsLib.builder.axes:Instance user-space location (400.0) redefined by 'Web'
WARNING:glyphsLib.builder.sources:Glyph(s) ['a'], brace layer '{155, 100}' defines more locations than there are design axes.
fontmake: Error: In 'master_ufo/GlyphsUnitTestSans-Light.ufo': Writing UFO source failed: 'fontinfo.plist' could not be written on <osfs '/var/folders/js/26vkgd8s5x15c5kwbhq2c2jw0000gn/T/tmp6vklkfae__tempfs__/GlyphsUnitTestSans-Light.ufo'> because the data is not properly formatted: unsupported type: <class 'glyphsLib.classes.GSMetricValue'>
```

**Comment by simoncozens** [comment]
Oh there are lots of problems. The unicode values are all bananas.

**Comment by simoncozens** [comment]
Holy macaroni, it works - I just built a variable font with fontmake from a Glyphs3 source!

There's that failing test, and I don't know how to fix it - it's probably something to do with mapping userspace/designspace locations. The axes stuff could be made a lot cleaner. It could be faster. But I'm leaving this for now.

**Comment by simoncozens** [comment]
The commits above add support for Glyphs 3 feature tokens.

**Comment by simoncozens** [comment]
Done. Ship it.

**Comment by simoncozens** [comment]
Sorry to do this but - just hold off for a minute. This is **shockingly** slow, and I think I *can* get it using the openstep parser after all.

**Comment by simoncozens** [comment]
OK, this now uses the C plist parser and it's super fast again.

**Comment by simoncozens** [comment]
Bump. This is also faster than master for Glyphs 2 files, by the way. And the longer we leave it the more likely we will have problems untangling conflicts.

**Comment by stone-zeng** [comment]
In Glyphs 3 the `weightClass` value now becomes an int rather than a string, so

```diff
diff --git a/venv/lib/python3.9/site-packages/glyphsLib/builder/axes.py b/venv/lib/python3.9/site-packages/glyphsLib/builder/axes.py
index 76a9c2b5..f268db09 100644
--- a/venv/lib/python3.9/site-packages/glyphsLib/builder/axes.py
+++ b/venv/lib/python3.9/site-packages/glyphsLib/builder/axes.py
@@ -77,7 +77,10 @@ def user_loc_string_to_value(axis_tag, user_loc):
         try:
             value = _nospace_lookup(WEIGHT_CODES, user_loc)
         except KeyError:
-            return None
+            try:
+                value = int(user_loc)
+            except ValueError:
+                return None
         return class_to_value("wght", value)
     elif axis_tag == "wdth":
         try:
```

**Comment by simoncozens** [comment]
This just recently got used to onboard a Google Fonts font, which shook out some problems.

**Comment by simoncozens** [comment]
Here (from babelfont) is the code to read a glyphspackage file. This should be fairly easy to adapt for glyphsLib once this PR is in.

```python
        infofile = os.path.join(self.filename, "fontinfo.plist")
        orderfile = os.path.join(self.filename, "order.plist")
        glyphorder = openstep_plist.load(open(orderfile, "r"))
        self.scratch["plist"] = openstep_plist.load(
            open(infofile, "r"), use_numbers=True
        )
        self.scratch["plist"]["glyphs"] = []
        for g in glyphorder:
            glyphname = userNameToFileName(g) + ".glyph"
            glyphfile = os.path.join(self.filename, "glyphs", glyphname)
            self.scratch["plist"]["glyphs"].append(
                openstep_plist.load(open(glyphfile, "r"), use_numbers=True)
            )

        return super()._load()
```

File: Lib/glyphsLib/writer.py [hunk 2]
```diff
@@ -140,8 +162,12 @@ def writeValue(self, value, forKey=None, forType=None):
...
         else:
             value = str(value)
-            if forKey != "unicode":
-                value = escape_string(value)
+            if self.format_version < 3:
+                if forKey != "unicode":
+                    value = escape_string(value)
+            else:
+                if _needs_quotes(value) or " " in value:
+                    value = '"%s"' % value
```
Comment (belluzj) [inline Lib/glyphsLib/writer.py:170]: What about quotes inside the value? e.g. value = '"'?

**Review by belluzj — COMMENTED** [review]
I have started reviewing this, there's a lot to take in. I still haven't read all the changes but I'm stopping for today. I have only a couple surface comments for now, but I'll come back to read everything, look more in depth at the changes and give the code a try on glyphs files that I have at hand. Thanks for working on that!

**Comment by madig** [comment]
Is there any chance this PR could be broken up into 1. parser update to openstep-plist and 2. adding Glyphs 3 support? It would reduce the review area.

**Comment by simoncozens** [comment]
Look at it the other way around: up to 4b661f8 is adding glyphs 3 support, and c56e6c4 onwards adds openstep-plist support.

**Comment by khaledhosny** [comment]
Does this PR affect version 2 support negatively? If not I think it should be merged, even of version 3 support might need further work and testing, nothing can be made perfect from the first time.

**Comment by madig** [comment]
To look for negative impacts, one has to review the PR and it's huge :grimacing:

**Comment by khaledhosny** [comment]
Given the lack of manpower, I’d take the heavy handed approach of testing this branch with a big collection of Glyphs 2 files, if nothing breaks then it is good and merge. Given that Noto is the original driving force behind glyphsLib, then probably that is a good collection to test with. If this is fine, I can try doing that.

**Comment by simoncozens** [comment]
To me the whole point of having a regression test suite is to give you confidence that new changes don’t introduce regressions in behaviour that you care about. If you don’t think the test suite is adequate and you don’t trust it to do that, by all means add new test cases.

**Comment by madig** [comment]
I have no idea if the existing test suite really covers everything that is relevant, so if someone can spend some time giving this PR a good thrashing with various Glyphs.app files, that'd help.

**Comment by khaledhosny** [comment]
I tried to build noto fonts, but like have of the fonts don’t even build with its current setup, so I’m giving up.

**Comment by madig** [comment]
You mean Noto does not build from source right now?

**Comment by khaledhosny** [comment]
Yes, the first font that fails is the N'ko font, but if you delete it other fonts will fail

**Comment by madig** [comment]
That's funny. Their CI seems to be unhinged. Maybe file an issue? https://github.com/googlefonts/noto-source/issues

**Comment by moyogo** [comment]
Running `build src/NotoSansNKo/NotoSansNKo.glyphs` works for me, with this PR as well.
@khaledhosny What commands did you run?

**Comment by khaledhosny** [comment]
I used `./build all`, it stopped with:
```
==== building src/NotoSansNKo/NotoSansNKo.plist ====
INFO:fontmake.font_project:Building master UFOs and designspace from Glyphs source
INFO:glyphsLib.classes:Parsing "src/NotoSansNKo/NotoSansNKo.glyphs" file into <GSFont>
INFO:fontmake.font_project:Building OTF for NotoSansNKo-Regular
INFO:ufo2ft:Pre-processing glyphs
INFO:ufo2ft.filters.base:Running DecomposeComponentsFilter on NotoSansNKo-Regular
INFO:ufo2ft.filters.base:Running RemoveOverlapsFilter on NotoSansNKo-Regular
INFO:ufo2ft:Building OpenType tables
fontmake: Error: In 'src/NotoSansNKo/NotoSansNKo.glyphs' -> 'master_ufo/NotoSansNKo-Regular.designspace' -> 'master_ufo/NotoSansNKo-Regular.ufo': Compiling UFO failed: (b'nko', 'str', 'ScriptRecord[0]', 'ScriptList')
INFO:fontmake.font_project:Building master UFOs and designspace from Glyphs source
INFO:glyphsLib.classes:Parsing "src/NotoSansNKo/NotoSansNKo.glyphs" file into <GSFont>
INFO:fontmake.font_project:Interpolating master UFOs from designspace
INFO:fontmake.font_project:Generating instance UFO for "Noto Sans NKo Regular"
INFO:fontmake.font_project:Building OTF for NotoSansNKo-Regular
INFO:ufo2ft:Pre-processing glyphs
INFO:ufo2ft.filters.base:Running DecomposeComponentsFilter on NotoSansNKo-Regular
INFO:ufo2ft.filters.base:Running RemoveOverlapsFilter on NotoSansNKo-Regular
INFO:ufo2ft:Building OpenType tables
INFO:ufo2ft.postProcessor:Subroutinizing CFF table with cffsubr
INFO:fontTools.varLib:Axes:
[{'default': 400.0,
  'hidden': False,
  'labelNames': {'en': 'Weight'},
  'map': [(400.0, 100.0)],
  'maximum': 400.0,
  'minimum': 400.0,
  'name': 'Weight',
  'tag': 'wght'}]
INFO:fontTools.varLib:Internal master locations:
[{'Weight': 100.0}]
INFO:fontTools.varLib:Internal axis supports:
{'Weight': [100.0, 100.0, 100.0]}
INFO:fontTools.varLib:Normalized master locations:
[{'Weight': 0.0}]
INFO:fontTools.varLib:Index of base master: 0
INFO:fontTools.varLib.interpolate_layout:Building interpolated font
INFO:fontTools.varLib.interpolate_layout:Loading master fonts
fontmake: Error: In 'src/NotoSansNKo/NotoSansNKo.glyphs' -> 'master_ufo/NotoSansNKo-Regular.designspace': Generating fonts from Designspace failed: Invalid master path: 'master_otf/NotoSansNKo-Regular.otf'
```

**Comment by moyogo** [comment]
Thanks @khaledhosny. I should have seen that.

**Comment by jenskutilek** [comment]
I have quite a collection of Glyphs 2 and 3 files, but most of them have TrueType hints, so glyphsLib bails out when trying to build them ...

**Comment by jenskutilek** [comment]
Oh, I’m in a special kind of Glyphs hell ;) In Sudo there are some TT hints that can not be seen in TT instructor nor deleted by setting `layer.hints = []`

**Comment by simoncozens** [comment]
Thanks for this. I deleted the hints by hand and it converted fine, so I will see what's going on with the hint parsing code.

**Comment by khaledhosny** [comment]
I gave up trying to build Noto and opted for something less ambitious, I changed the build script to output UFO from fontmake instead and compared the output with glyphsLib main and this branch.

I’m getting this warning with all fonts:
```
glyphsLib/classes.py:1707: UserWarning: GSFontMaster metric ascender accessed without an associated font!
  warnings.warn(
glyphsLib/classes.py:1707: UserWarning: GSFontMaster metric cap height accessed without an associated font!
  warnings.warn(
glyphsLib/classes.py:1707: UserWarning: GSFontMaster metric descender accessed without an associated font!
  warnings.warn(
glyphsLib/classes.py:1707: UserWarning: GSFontMaster metric x-height accessed without an associated font!
  warnings.warn(
```

 Here are some of the common diff’s I’m seeing:
```diff
diff --git a/master_ufo/Cousine-Bold.ufo/lib.plist b/master_ufo/Cousine-Bold.ufo/lib.plist
index 2582c29164..342aec4533 100644
--- a/master_ufo/Cousine-Bold.ufo/lib.plist
+++ b/master_ufo/Cousine-Bold.ufo/lib.plist
@@ -18,7 +18,7 @@
     <key>com.schriftgestaltung.appVersion</key>
     <string>1277</string>
     <key>com.schriftgestaltung.customParameter.GSFont.disablesAutomaticAlignment</key>
-    <true/>
+    <integer>1</integer>
     <key>com.schriftgestaltung.customParameter.GSFont.useNiceNames</key>
     <integer>0</integer>
     <key>com.schriftgestaltung.customParameter.GSFontMaster.customValue</key>
```

```diff
diff --git a/master_ufo/NotoKufiArabic-Black.ufo/lib.plist b/master_ufo/NotoKufiArabic-Black.ufo/lib.plist
index bacb012519..c270714eae 100644
--- a/master_ufo/NotoKufiArabic-Black.ufo/lib.plist
+++ b/master_ufo/NotoKufiArabic-Black.ufo/lib.plist
@@ -17,15 +17,6 @@
     <string>1350</string>
     <key>com.schriftgestaltung.customName</key>
     <string>Black</string>
-    <key>com.schriftgestaltung.customParameter.GSFont.Axes</key>
-    <array>
-      <dict>
-        <key>Name</key>
-        <string>Weight</string>
-        <key>Tag</key>
-        <string>wght</string>
-      </dict>
-    </array>
     <key>com.schriftgestaltung.customParameter.GSFont.Enforce Compatibility Check</key>
     <integer>1</integer>
     <key>com.schriftgestaltung.customParameter.GSFont.disablesAutomaticAlignment</key>
```

```diff
diff --git a/master_ufo/NotoLoopedLao-Regular.ufo/glyphs/aV_owel-lao.glif b/master_ufo/NotoLoopedLao-Regular.ufo/glyphs/aV_owel-lao.glif
index 673edf0b65..946f734f3e 100644
--- a/master_ufo/NotoLoopedLao-Regular.ufo/glyphs/aV_owel-lao.glif
+++ b/master_ufo/NotoLoopedLao-Regular.ufo/glyphs/aV_owel-lao.glif
@@ -96,7 +96,7 @@
       <array>
         <dict>
           <key>horizontal</key>
-          <true/>
+          <integer>1</integer>
           <key>options</key>
           <integer>0</integer>
           <key>origin</key>
@@ -116,7 +116,7 @@
         </dict>
         <dict>
           <key>horizontal</key>
-          <true/>
+          <integer>1</integer>
           <key>options</key>
           <integer>0</integer>
           <key>origin</key>
```

```diff
diff --git a/master_ufo/NotoSans-Condensed.ufo/glyphs/ostroke-blackletter.glif b/master_ufo/NotoSans-Condensed.ufo/glyphs/ostroke-blackletter.glif
index 79d52763b7..3d0c8b559d 100644
--- a/master_ufo/NotoSans-Condensed.ufo/glyphs/ostroke-blackletter.glif
+++ b/master_ufo/NotoSans-Condensed.ufo/glyphs/ostroke-blackletter.glif
@@ -34,7 +34,7 @@
       <string>Lowercase</string>
       <key>com.schriftgestaltung.componentsLocked</key>
       <array>
-        <true/>
+        <integer>1</integer>
       </array>
     </dict>
   </lib>
```

(So far, this true/1 change accounts for most of the diff)

**Comment by khaledhosny** [comment]
```diff
diff --git a/master_ufo/NotoSansIV-Italic.designspace b/master_ufo/NotoSansIV-Italic.designspace
index 7fd075a26d..6911400c9c 100644
--- a/master_ufo/NotoSansIV-Italic.designspace
+++ b/master_ufo/NotoSansIV-Italic.designspace
@@ -4,13 +4,11 @@
     <axis tag="wght" name="Weight" minimum="100" maximum="100" default="100">
       <map input="100" output="26"/>
     </axis>
-    <axis tag="wdth" name="Width" minimum="100" maximum="100" default="100"/>
   </axes>
   <sources>
     <source filename="NotoSansIV-LightItalic.ufo" name="Noto Sans IV Light Italic" familyname="Noto Sans IV" stylename="Light Italic">
       <location>
         <dimension name="Weight" xvalue="26"/>
-        <dimension name="Width" xvalue="100"/>
       </location>
     </source>
     <source filename="NotoSansIV-Italic.ufo" name="Noto Sans IV Italic" familyname="Noto Sans IV" stylename="Italic">
@@ -20,43 +18,36 @@
       <info copy="1"/>
       <location>
         <dimension name="Weight" xvalue="90"/>
-        <dimension name="Width" xvalue="100"/>
       </location>
     </source>
     <source filename="NotoSansIV-SemiBoldItalic.ufo" name="Noto Sans IV SemiBold Italic" familyname="Noto Sans IV" stylename="SemiBold Italic">
       <location>
         <dimension name="Weight" xvalue="151"/>
-        <dimension name="Width" xvalue="100"/>
       </location>
     </source>
     <source filename="NotoSansIV-BoldItalic.ufo" name="Noto Sans IV Bold Italic" familyname="Noto Sans IV" stylename="Bold Italic">
       <location>
         <dimension name="Weight" xvalue="190"/>
-        <dimension name="Width" xvalue="100"/>
       </location>
     </source>
     <source filename="NotoSansIV-CondensedLightItalic.ufo" name="Noto Sans IV Condensed Light Italic" familyname="Noto Sans IV" stylename="Condensed Light Italic">
       <location>
         <dimension name="Weight" xvalue="26"/>
-        <dimension name="Width" xvalue="70"/>
       </location>
     </source>
     <source filename="NotoSansIV-CondensedItalic.ufo" name="Noto Sans IV Condensed Italic" familyname="Noto Sans IV" stylename="Condensed Italic">
       <location>
         <dimension name="Weight" xvalue="90"/>
-        <dimension name="Width" xvalue="70"/>
       </location>
     </source>
     <source filename="NotoSansIV-CondensedSemiBoldItalic.ufo" name="Noto Sans IV Condensed SemiBold Italic" familyname="Noto Sans IV" stylename="Condensed SemiBold Italic">
       <location>
         <dimension name="Weight" xvalue="151"/>
-        <dimension name="Width" xvalue="70"/>
       </location>
     </source>
     <source filename="NotoSansIV-CondensedBoldItalic.ufo" name="Noto Sans IV Condensed Bold Italic" familyname="Noto Sans IV" stylename="Condensed Bold Italic">
       <location>
         <dimension name="Weight" xvalue="190"/>
-        <dimension name="Width" xvalue="70"/>
       </location>
     </source>
   </sources>
@@ -64,7 +55,6 @@
     <instance name="Noto Sans IV Thin Italic" familyname="Noto Sans IV" stylename="Thin Italic" filename="../instance_ufo/NotoSansIV-ThinItalic.ufo" stylemapfamilyname="Noto Sans IV Thin" stylemapstylename="italic">
       <location>
         <dimension name="Weight" xvalue="26"/>
-        <dimension name="Width" xvalue="100"/>
       </location>
       <kerning/>
       <info/>
```

**Comment by khaledhosny** [comment]
```diff
diff --git a/master_ufo/NotoTraditionalNushu.designspace b/master_ufo/NotoTraditionalNushu.designspace
index 4441f86bf9..8fe320d152 100644
--- a/master_ufo/NotoTraditionalNushu.designspace
+++ b/master_ufo/NotoTraditionalNushu.designspace
@@ -2,9 +2,7 @@
 <designspace format="4.1">
   <axes>
     <axis tag="wght" name="Weight" minimum="300" maximum="700" default="400"/>
-    <axis tag="wdth" name="Width" minimum="100" maximum="100" default="100">
-      <map input="100" output="5"/>
-    </axis>
+    <axis tag="wdth" name="Width" minimum="5" maximum="5" default="5"/>
   </axes>
   <sources>
     <source filename="NotoTraditionalNushu-Light.ufo" name="Noto Traditional Nushu Light" familyname="Noto Traditional Nushu" stylename="Light">
```

The rest of the diffs are instances of the examples above.

**Comment by simoncozens** [comment]
5102f8a might fix the diffs you saw above.

**Comment by simoncozens** [comment]
-> #691

**Pull request closed without merging**